crates/schema/src/type_for_generate.rs

```diff
@@ -0,0 +1,516 @@
+
+use crate::{
+    error::{IdentifierError, PrettyAlgebraicType},
+    identifier::Identifier,
+};
+
+/// Errors that can occur when rearranging types for client codegen.
+#[derive(thiserror::Error, Debug, PartialOrd, Ord, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ClientCodegenError {
+    #[error(
+        "internal codegen error: non-special product or sum type {ty} cannot be used to generate a client type use"
+    )]
+    NonSpecialTypeNotAUse { ty: PrettyAlgebraicType },
+
+    #[error("internal codegen error: invalid AlgebraicTypeRef")]
+    TypeRefError(#[from] TypeRefError),
+
+    #[error("internal codegen error: type ref {ref_} was not pre-declared as a definition")]
+    NonDeclaredTypeDef { ref_: AlgebraicTypeRef },
+
+    #[error("internal codegen error: all type elements require names: {ty}")]
+    NamelessTypeDef { ty: PrettyAlgebraicType },
+
+    #[error("internal codegen error: type {ty} is not valid for generating a definition")]
+    NotValidForDefinition { ty: PrettyAlgebraicType },
+
+    #[error("type {ty} contains identifier error {err}")]
+    NotValidIdentifier {
+        ty: PrettyAlgebraicType,
+        err: IdentifierError,
+    },
+}
+
+type Result<T> = std::result::Result<T, ErrorStream<ClientCodegenError>>;
+
+/// A typespace for generating client code.
+///
+/// The key difference is that this typespace stores only `AlgebraicTypeDef`s, not `AlgebraicType`s.
+/// We use the same `AlgebraicTypeRef`s from the original typespace.
+/// The difference is that `AlgebraicTypeRef`s ONLY point to `AlgebraicTypeDef`s.
+/// Chains of `AlgebraicTypeRef`s in the original `Typespace` are contracted to point to their ending `AlgebraicTypeDef`.
+#[derive(Debug, Clone)]
+pub struct TypespaceForGenerate {
+    defs: HashMap<AlgebraicTypeRef, AlgebraicTypeDef>,
+}
+
+impl TypespaceForGenerate {
+    /// Get the definitions of the typespace.
+    pub fn defs(&self) -> &HashMap<AlgebraicTypeRef, AlgebraicTypeDef> {
+        &self.defs
+    }
+
+    /// Build a `TypespaceForGenerate`.
+    ///
+    /// We're required to declare known definitions up front.
+    /// This is required for distinguishing between a use of the unit type, and a reference to a type declaration of a product type with no elements.
+    pub fn builder(
+        typespace: &Typespace,
+        known_definitions: impl IntoIterator<Item = AlgebraicTypeRef>,
+    ) -> TypespaceForGenerateBuilder<'_> {
+        TypespaceForGenerateBuilder {
+            typespace,
+            result: TypespaceForGenerate { defs: HashMap::new() },
+            known_definitions: known_definitions.into_iter().collect(),
+            uses: HashSet::new(),
+            known_uses: HashMap::new(),
+            currently_touching: HashSet::new(),
+        }
+    }
+}
+
+/// An algebraic type definition.
+#[derive(Debug, Clone)]
+pub enum AlgebraicTypeDef {
+    /// A product type declaration.
+    Product(ProductTypeDef),
+    /// A sum type declaration.
+    Sum(SumTypeDef),
```

Comment:
Can we have a separate variant for plain enum? They already need to be handled as a different type in languages without tagged enum support, and I imagine we might need that even more in the future.

Reply:
👍